File: Backend/node/acciones_bd.js
const mysql = require('mysql2');

// Configuración de la conexión a la base de datos
const config = {
    host: 'localhost',
    user: 'root',
    password: '123456789',  // Cambia esto por tu contraseña de MySQL
    database: 'taskflow_cloud'
};

// Crear la conexión
const conn = mysql.createPool(config).promise();

// Función para logearse
async function logearme(username, password) {
    try {
        console.log(username, password, 'logearme');
        const [rows] = await conn.execute(
            `SELECT id FROM usuarios WHERE nombre_usuario = ? AND contrasena = SHA2(?, 256) LIMIT 1`,
            [username, password]
        );

        if (rows.length === 0) return false; // Usuario no encontrado

        const userId = rows[0].id;

        // Verificar si ya hay un usuario logeado
        const [logged] = await conn.execute(`SELECT usuario_id FROM userlogeado WHERE id = 1`);
        if (logged.length > 0) {
            if (logged[0].usuario_id === userId) return null; // Ya está logeado
            await conn.execute(`UPDATE userlogeado SET usuario_id = ? WHERE id = 1`, [userId]);
        } else {
            await conn.execute(`INSERT INTO userlogeado (id, usuario_id) VALUES (1, ?)`, [userId]);
        }

        return true;
    } catch (error) {
        console.error('Error en logearme:', error);
        return false;
    }
}

// Función para desloguearse
async function desloguear() {
    try {
        const [result] = await conn.execute(`DELETE FROM userlogeado WHERE id = 1`);
        return result.affectedRows > 0;
    } catch (error) {
        console.error('Error en desloguear:', error);
        return false;
    }
}

// Función para registrar usuario
async function registrarusuario(username, email, password, imagen) {
    try {
        // Verificar si el usuario ya existe
        const [exist] = await conn.execute(`SELECT COUNT(*) AS count FROM usuarios WHERE nombre_usuario = ?`, [username]);
        if (exist[0].count > 0) return false;

        // Registrar usuario
        await conn.execute(
            `INSERT INTO usuarios (nombre_usuario, correo, contrasena, imagen_perfil_url) VALUES (?, ?, SHA2(?, 256), ?)`,
            [username, email, password, imagen]
        );
        return true;
    } catch (error) {
        console.error('Error en registrarusuario:', error);
        return false;
    }
}


// Función para obtener la información del usuario logeado
async function obtenerUser() {
    try {
        // Verificar si hay un usuario logeado
        const [logged] = await conn.execute(`SELECT usuario_id FROM userlogeado WHERE id = 1`);
        if (logged.length === 0) {
            return null; // No hay usuario logeado
        }

        const userId = logged[0].usuario_id;

        // Obtener información del usuario logeado desde la tabla usuarios
        const [user] = await conn.execute(`SELECT id, nombre_usuario, correo, imagen_perfil_url FROM usuarios WHERE id = ?`, [userId]);

        if (user.length === 0) {
            return null; // No se encontró el usuario
        }

        // Retornar la información del usuario
        return user[0];
    } catch (error) {
        console.error('Error al obtener usuario logeado:', error);
        return null;
    }
}


async function cargarArchivo(usuario_id, nombre_archivo, tipo_archivo, archivo_url) {
    try {
        await conn.execute(
            `INSERT INTO archivos (usuario_id, nombre_archivo, tipo_archivo, url_archivo) 
            VALUES (?, ?, ?, ?)`,
            [usuario_id, nombre_archivo, tipo_archivo, archivo_url]
        );
        return true;
    } catch (error) {
        console.error('Error al cargar archivo:', error);
        return false;
    }
}

async function listarArchivos(usuarioId) {
    try {
        const [archivos] = await conn.execute(
            `SELECT * FROM archivos WHERE usuario_id = ?`,
            [usuarioId]
        );
        return archivos;
    } catch (error) {
        console.error('Error al listar archivos:', error);
        return [];
    }
}


async function obtenerUsuarioLogeado() {
    try {
        const [result] = await conn.execute(
            `SELECT usuario_id FROM userlogeado WHERE id = 1`
        );
        return result[0]?.usuario_id || null;
    } catch (error) {
        console.error('Error al obtener usuario logeado:', error);
        return null;
    }
}



// Crear tarea
function crearTarea(usuario_id, titulo, descripcion, fecha_creacion) {
    return new Promise((resolve, reject) => {
      const consulta = `
        INSERT INTO tareas (usuario_id, titulo, descripcion, fecha_creacion) 
        VALUES (?, ?, ?, ?)
      `;
      // Cambiar 'conexion' por 'conn' para utilizar la conexión correctamente
      conn.execute(consulta, [usuario_id, titulo, descripcion, fecha_creacion])
        .then(() => resolve(true))
        .catch(err => {
          console.log('Error al crear tarea:', err);
          reject(false);
        });
    });
  }
  

// Editar tarea
async function editarTarea(tarea_id, titulo, descripcion) {
    try {
        const conexion = await conn.getConnection();
        let consulta = 'UPDATE tareas SET ';
        let valores = [];
        
        if (titulo) {
            consulta += 'titulo = ?, ';
            valores.push(titulo);
        }
        if (descripcion) {
            consulta += 'descripcion = ?, ';
            valores.push(descripcion);
        }
        if (valores.length === 0) return false;

        consulta = consulta.slice(0, -2) + ' WHERE id = ?';
        valores.push(tarea_id);

        const [resultado] = await conexion.execute(consulta, valores);
        conexion.release();
        return resultado.affectedRows > 0;
    } catch (error) {
        console.error('Error al editar tarea:', error);
        return false;
    }
}

// Marcar tarea como completada
async function completarTarea(tarea_id) {
    try {
        const conexion = await conn.getConnection();
        const [resultado] = await conexion.execute(
            `UPDATE tareas SET completada = TRUE WHERE id = ?`,
            [tarea_id]
        );
        conexion.release();
        return resultado.affectedRows > 0;
    } catch (error) {
        console.error('Error al completar tarea:', error);
        return false;
    }
}

// Eliminar tarea
async function eliminarTarea(tarea_id) {
    try {
        const conexion = await conn.getConnection();
        const [resultado] = await conexion.execute(
            `DELETE FROM tareas WHERE id = ?`,
            [tarea_id]
        );
        conexion.release();
        return resultado.affectedRows > 0;
    } catch (error) {
        console.error('Error al eliminar tarea:', error);
        return false;
    }
}

async function listar_tareas(usuarioId) {
    try {
        const [tareas] = await conn.execute(
            `SELECT * FROM tareas WHERE usuario_id = ?`,
            [usuarioId]
        );
        
        // Formatear la fecha como dd/mm/aaaa
        tareas.forEach(tarea => {
            if (tarea.fecha_creacion) {
                const fecha = new Date(tarea.fecha_creacion);
                const dia = String(fecha.getDate()).padStart(2, '0');
                const mes = String(fecha.getMonth() + 1).padStart(2, '0');
                const anio = fecha.getFullYear();
                tarea.fecha_creacion = `${dia}/${mes}/${anio}`;
            }
        });
        return tareas;
    } catch (error) {
        console.error('Error al listar tareas:', error);
        return [];
    }
}
async function descompletar_tarea(tarea_id) {
    try {
        const conexion = await conn.getConnection();
        const [resultado] = await conexion.execute(
            `UPDATE tareas SET completada = FALSE WHERE id = ?`,
            [tarea_id]
        );
        conexion.release();
        return resultado.affectedRows > 0;
    } catch (error) {
        console.error('Error al completar tarea:', error);
        return false;
    }
}

module.exports = {
    logearme, desloguear, registrarusuario,
    cargarArchivo, listarArchivos, obtenerUsuarioLogeado,
    crearTarea, editarTarea,
    completarTarea, eliminarTarea, obtenerUser,listar_tareas,descompletar_tarea
};

File: Backend/python/acciones_bd.py
import mysql.connector
from datetime import datetime

# Configuración de la conexión a la base de datos
config = {
    'user': 'root',         
    'password': '123456789',  
    'host': 'localhost',
    'database': 'taskflow_cloud'
}

def get_connection():
    """Establece y retorna una conexión a la base de datos."""
    return mysql.connector.connect(**config)


def obtener_usuario_logeado():
    try:
        conn = get_connection()
        cursor = conn.cursor()

        cursor.execute("SELECT usuario_id FROM userlogeado WHERE id = 1;")
        result = cursor.fetchone()

        cursor.close()
        conn.close()

        return result[0] if result else None

    except Exception as e:
        print("Error al obtener usuario logeado:", e)
        return None
    

def logearme(username, password):
    try:
        conn = get_connection()
        cursor = conn.cursor()

        query = """
            SELECT id 
            FROM usuarios 
            WHERE nombre_usuario = %s 
              AND contrasena = SHA2(%s, 256)
            LIMIT 1;
        """
        cursor.execute(query, (username, password))
        result = cursor.fetchone()
        if not result:
            return False

        user_id = result[0]

        cursor.execute("SELECT usuario_id FROM userlogeado WHERE id = 1;")
        registro = cursor.fetchone()
        if registro:
            if registro[0] == user_id:
                return None
            else:
                cursor.execute("UPDATE userlogeado SET usuario_id = %s WHERE id = 1;", (user_id,))
                conn.commit()
        else:
            cursor.execute("INSERT INTO userlogeado (id, usuario_id) VALUES (1, %s);", (user_id,))
            conn.commit()

        cursor.close()
        conn.close()
        return True

    except Exception as e:
        print("Error al logear:", e)
        return False

def desloguear():
    try:
        conexion = get_connection()
        cursor = conexion.cursor()

        cursor.execute("DELETE FROM userlogeado WHERE id = 1;")
        filas_afectadas = cursor.rowcount

        if filas_afectadas == 0:
            cursor.close()
            conexion.close()
            return False

        conexion.commit()
        cursor.close()
        conexion.close()
        return True

    except Exception as e:
        print("Error al desloguear:", e)
        if conexion:
            conexion.rollback()
        return False

def crear_tarea(usuario_id, titulo, descripcion, fecha_creacion):
    try:
        conexion = get_connection()
        ejecutar = conexion.cursor()
        
        consulta = """
            INSERT INTO tareas (usuario_id, titulo, descripcion, fecha_creacion) 
            VALUES (%s, %s, %s, %s)
        """
        ejecutar.execute(consulta, (usuario_id, titulo, descripcion, fecha_creacion))
        conexion.commit()
        ejecutar.close()
        conexion.close()
        return True
    except Exception as e:
        print(f"Error al crear tarea: {e}")
        return False

    

def editar_tarea(tarea_id, titulo=None, descripcion=None):
    try:
        conexion = get_connection()
        ejecutar = conexion.cursor()

        valores = []
        consulta = "UPDATE tareas SET "

        if titulo is not None:
            consulta += "titulo = %s, "
            valores.append(titulo)

        if descripcion is not None:
            consulta += "descripcion = %s, "
            valores.append(descripcion)

        if not valores:
            print("No hay valores para actualizar")
            return False  

        consulta = consulta.rstrip(", ") + " WHERE id = %s"
        valores.append(tarea_id)

        ejecutar.execute(consulta, tuple(valores))
        conexion.commit()

        filas_afectadas = ejecutar.rowcount

        ejecutar.close()
        conexion.close()

        return filas_afectadas > 0
    except Exception as e:
        print(f"Error al editar tarea: {e}")
        return False

def registrarusuario(username, email, password, imagen):
    try:
        conexion = get_connection()
        ejecutar = conexion.cursor()

        consulta_verificacion = "SELECT COUNT(*) FROM usuarios WHERE nombre_usuario = %s"
        ejecutar.execute(consulta_verificacion, (username,))
        existe_usuario = ejecutar.fetchone()[0]

        if existe_usuario > 0:
            ejecutar.close()
            conexion.close()
            return False

        consulta = """
            INSERT INTO usuarios (nombre_usuario, correo, contrasena, imagen_perfil_url) 
            VALUES (%s, %s, SHA2(%s, 256), %s)
        """
        
        ejecutar.execute(consulta, (username, email, password, imagen))
        conexion.commit()

        ejecutar.close()
        conexion.close()
        return True

    except Exception as e:
        print(f"Error al registrar usuario: {e}")
        if conexion:
            conexion.rollback()
        return False

# Cargar archivo
def cargar_archivo(usuario_id, nombre_archivo, tipo_archivo, archivo_url):
    try:
        conexion = get_connection()
        ejecutar = conexion.cursor()

        consulta = """
            INSERT INTO archivos (usuario_id, nombre_archivo, tipo_archivo, url_archivo) 
            VALUES (%s, %s, %s, %s)
        """
        ejecutar.execute(consulta, (usuario_id, nombre_archivo, tipo_archivo, archivo_url))
        conexion.commit()

        ejecutar.close()
        conexion.close()
        return True

    except Exception as e:
        print(f"Error al cargar archivo: {e}")
        if conexion:
            conexion.rollback()
        return False

# Listar archivos
def listar_archivos(usuario_id):
    try:
        conexion = get_connection()
        ejecutar = conexion.cursor()

        consulta = "SELECT * FROM archivos WHERE usuario_id = %s"
        ejecutar.execute(consulta, (usuario_id,))
        archivos = ejecutar.fetchall()

        ejecutar.close()
        conexion.close()

        return archivos if archivos else []

    except Exception as e:
        print(f"Error al listar archivos: {e}")
        return []
    
# Marcar tarea como completada
def completar_tarea(tarea_id):
    try:
        conexion = get_connection()
        ejecutar = conexion.cursor()
        consulta = "UPDATE tareas SET completada = TRUE WHERE id = %s"
        ejecutar.execute(consulta, (tarea_id,))
        conexion.commit()
        filas_afectadas = ejecutar.rowcount
        ejecutar.close()
        conexion.close()
        return filas_afectadas > 0
    except Exception as e:
        print(f"Error al completar tarea: {e}")
        if conexion:
            conexion.rollback()
        return False
# Marcar tarea como pendiente (descompletar)
def descompletar_tarea(tarea_id):
    try:
        conexion = get_connection()
        ejecutar = conexion.cursor()
        consulta = "UPDATE tareas SET completada = FALSE WHERE id = %s"
        ejecutar.execute(consulta, (tarea_id,))
        conexion.commit()
        filas_afectadas = ejecutar.rowcount
        ejecutar.close()
        conexion.close()
        return filas_afectadas > 0
    except Exception as e:
        print(f"Error al descompletar tarea: {e}")
        if conexion:
            conexion.rollback()
        return False



# Eliminar tarea
def eliminar_tarea(tarea_id):
    print(tarea_id)
    try:
        conexion = get_connection()
        ejecutar = conexion.cursor()
        consulta = "DELETE FROM tareas WHERE id = %s"
        ejecutar.execute(consulta, (tarea_id,))
        conexion.commit()
        filas_afectadas = ejecutar.rowcount
        ejecutar.close()
        conexion.close()
        return filas_afectadas > 0
    except Exception as e:
        print(f"Error al eliminar tarea: {e}")
        if conexion:
            conexion.rollback()
        return False


def obtenerinfologeado():
    try:
        conn = get_connection()
        cursor = conn.cursor()

        # Verificar si hay un usuario logeado en la tabla userlogeado
        cursor.execute("SELECT usuario_id FROM userlogeado WHERE id = 1;")
        registro = cursor.fetchone()
        
        if not registro:
            # No hay usuario logeado
            print("No hay usuario logeado.")
            return None

        user_id = registro[0]

        # Obtener toda la información del usuario logeado desde la tabla usuarios
        cursor.execute("SELECT id, nombre_usuario, correo, imagen_perfil_url FROM usuarios WHERE id = %s;", (user_id,))
        usuario_info = cursor.fetchone()

        if not usuario_info:
            print("No se encontró la información del usuario.")
            return None

        cursor.close()
        conn.close()
        return {
            "id": usuario_info[0],
            "nombre_usuario": usuario_info[1],
            "correo": usuario_info[2],
            "imagen_perfil_url": usuario_info[3]
        }

    except Exception as e:
        print("Error al obtener la información del usuario logeado:", e)
        return None

# Función para listar tareas con formato de fecha dd/mm/aaaa
def listar_tareas(usuario_id):
    try:
        conexion = get_connection()
        ejecutar = conexion.cursor(dictionary=True)  

        consulta = "SELECT * FROM tareas WHERE usuario_id = %s"
        ejecutar.execute(consulta, (usuario_id,))
        tareas = ejecutar.fetchall()

        ejecutar.close()
        conexion.close()

        # Formatear la fecha antes de devolver los resultados
        for tarea in tareas:
            if 'fecha_creacion' in tarea and tarea['fecha_creacion']:
                tarea['fecha_creacion'] = datetime.strptime(str(tarea['fecha_creacion']), "%Y-%m-%d %H:%M:%S").strftime("%d/%m/%Y")

        return tareas if tareas else []

    except Exception as e:
        print(f"Error al listar tareas: {e}")
        return []
